test/LibraryManager.IntegrationTest/LibmanTestsUtility.cs

```diff
@@ -8,6 +9,164 @@ namespace Microsoft.Web.LibraryManager.IntegrationTest
 {
     public class LibmanTestsUtility
     {
+        private static HashSet<string> GetTopLevelDirectoriesAndFiles(string cwd, bool caseInsensitive)
+        {
+            HashSet<string> topLevelItems = caseInsensitive ? new HashSet<string>(StringComparer.OrdinalIgnoreCase) : new HashSet<string>();
+            var dir = new DirectoryInfo(cwd);
+
+            if(dir.Exists)
+            {
+                foreach(FileSystemInfo item in dir.EnumerateDirectories())
+                {
+                    topLevelItems.Add(item.Name);
+                }
+                foreach(FileSystemInfo item in dir.EnumerateFiles())
+                {
+                    topLevelItems.Add(item.Name);
+                }
+            }
+
+            return topLevelItems;
+        }
+
+        public static void WaitForRestoredFiles(string cwd, IEnumerable<string> expectedFilesAndFolders, bool caseInsensitive, int timeout = 10000)
+        {
+            string errorMessage = WaitForRestoredFilesHelper(cwd, expectedFilesAndFolders, caseInsensitive, timeout);
+
+            if (errorMessage != null)
+            {
+                string newErrorMessage = WaitForRestoredFilesHelper(cwd, expectedFilesAndFolders, caseInsensitive, timeout);
+
+                if (newErrorMessage != null)
+                {
+                    errorMessage = String.Concat(errorMessage, "\r\nFailed even when forcing completion with double timeout");
+                }
+                else
+                {
+                    errorMessage = String.Concat(errorMessage, "\r\n*Didn't* fail when forcing completion with double timeout");
+                }
+
+                throw new TimeoutException(errorMessage);
+            }
+        }
+
+        public static void WaitForRestoredFile(string cwd, string expectedFileOrFolder, bool caseInsensitive, int timeout = 10000)
+        {
+            WaitForRestoredFiles(cwd, new[] { expectedFileOrFolder }, caseInsensitive, timeout);
+        }
+
+        public static void WaitForRestoredFileNotPresent(string cwd, string expectedFileOrFolder, bool caseInsensitive, int timeout = 10000)
+        {
+            string errorMessage = WaitForRestoredFilesHelper(cwd, new[] { expectedFileOrFolder }, caseInsensitive, timeout);
+
+            if (errorMessage == null)
```

Comment:
This is similar to the completion issue in that you appear to be waiting for a timeout for this method to pass

Reply:
Yes. But it will have different passing parameters and different message.

---
In reply to: [194397045](https://github.com/aspnet/LibraryManager/pull/113#discussion_r194397045) [](ancestors = 194397045)

Reply:
I see. This can be optimized.

---
In reply to: [194397045](https://github.com/aspnet/LibraryManager/pull/113#discussion_r194397045) [](ancestors = 194397045)

Reply:
This would be an different issue. Because the restoring is in progress after save the libman.json file. It needs sufficient time for restoring before we believe the unexpected file is not presented. 

---
In reply to: [194397045](https://github.com/aspnet/LibraryManager/pull/113#discussion_r194397045) [](ancestors = 194397045)

Reply:
This function is deleted.

---
In reply to: [194937801](https://github.com/aspnet/LibraryManager/pull/113#discussion_r194937801) [](ancestors = 194937801,194397045)